doc/icache.rst

```diff
@@ -0,0 +1,130 @@
+|                         |           | lines. This gives improved performance in     |
+|                         |           | systems with moderate latency by not          |
+|                         |           | polluting the cache with data that can be     |
+|                         |           | prefetched instead.                           |
++-------------------------+-----------+-----------------------------------------------+
+
+RAM Arrangement
+---------------
+
+The data RAMs are arranged as a series of 32 bit banks, the number depending on the cache line width and number of ways.
+Each group of banks forming a single way can be combined into wider RAM instances if required since they are always accessed together.
+
+Indicative RAM sizes for common configurations are given in the table below:
+
++-------------------------+-----------------+-----------------+
+| Cache config            | Tag RAMs        | Data RAMs       |
++=========================+=================+=================+
+| 4kB, 2 way, 64bit line  | 2 x 256 x 21bit | 4 x 256 x 32bit |
++-------------------------+-----------------+-----------------+
+| 4kB, 2 way, 128bit line | 2 x 128 x 21bit | 8 x 128 x 32bit |
++-------------------------+-----------------+-----------------+
+| 4kB, 4 way, 64bit line  | 4 x 128 x 21bit | 8 x 128 x 32bit |
++-------------------------+-----------------+-----------------+
+
+Sub Unit Description
+--------------------
+
+.. figure:: images/icache_block.svg
+   :name: icache_block
+   :align: center
+
+   Instruction Cache Block Diagram
+
+Prefetch Address
+^^^^^^^^^^^^^^^^
+
+The prefetch address is updated to the branch target on every branch.
+This address is then updated in cache-line increments each time a cache lookup is issued to the cache pipeline.
+
+Cache Pipeline
+^^^^^^^^^^^^^^
+
+The cache pipeline consists of two stages, IC0 and IC1.
+
+In IC0, lookup requests are arbitrated against cache allocation requests.
+Lookup requests have fixed priority since they naturally throttle themselves as fill buffer resources run out.
+The arbitrated request is made to the RAMs in IC0.
+
+In IC1, data from the RAMs are available and the cache hit status is determined.
+Hit data is multiplexed from the data RAMs based on the hitting way.
+If there was a cache miss, the victim way is chosen pseudo-randomly using a counter.
+
+Fill buffers
+^^^^^^^^^^^^
+
+The fill buffers perform several functions in the I$ and constitute most of it's complexity.
+
+- Since external requests can be made speculatively in parallel with the cache lookup, a fill buffer must be allocated in IC0 to track the request.
+- A fill buffer makes multiple external requests to memory to fetch the required data to fill a cache line (tracked via ``fill_ext_cnt_q``).
+- Returning data is tracked via ``fill_rvd_cnt_q``. Not all requests will fetch all their data, since requests can be cancelled due to a cache hit or an intervening branch.
```

Comment:
OK, sounds good. A minor comment: The sentence starting with "Not all requests..." should be moved to a new line (it will still be in the same list item).